cpp/src/arrow/compute/kernels/scalar_arithmetic.cc

```diff
@@ -2428,6 +2435,20 @@ void RegisterScalarArithmetic(FunctionRegistry* registry) {
     DCHECK_OK(subtract->AddKernel({in_type, in_type}, duration(unit), std::move(exec)));
   }
 
+  // Add subtract(time32, time32) -> duration
+  for (auto unit : {TimeUnit::SECOND, TimeUnit::MILLI}) {
+    InputType in_type(match::Time32TypeUnit(unit));
+    auto exec = ScalarBinaryEqualTypes<Int64Type, Time32Type, SubtractTime32>::Exec;
```

Comment:
I'm curious, why doesn't this simply work with `Subtract` as below?

Reply:
Because `Subtract` call we go through asserts Arg0 and Arg1 to be equal to T:
```
  template <typename T, typename Arg0, typename Arg1>
  static constexpr enable_if_signed_integer_value<T> Call(KernelContext*, Arg0 left,
                                                          Arg1 right, Status*) {
    static_assert(std::is_same<T, Arg0>::value && std::is_same<T, Arg1>::value, "");
    return arrow::internal::SafeSignedSubtract(left, right);
  }
```
Similar holds for date32. We could make `enable_if_signed_integer_value` accept Arg0 and T and make `enable_if_heterogeneous_signed_integer_value` to cover casting cases.

Reply:
Or we can simply remove the assert. It doesn't seem very useful TBH.

Reply:
Great to have that option :).

Reply:
Removed.